### PR TITLE
Attempt at fixing Sum type schema derivation for good

### LINKED ISSUE #1992: Scala 3 schema derivation duplicate types in union
Given the following:
```scala
sealed trait Parent
object Parent {
  implicit val schema: Schema[Any, Parent] = Schema.gen

  sealed trait OtherParent extends Parent
  object OtherParent {
    implicit val schema: Schema[Any, OtherParent] = Schema.gen
  }

  case class Child2(i: Int) extends Parent with OtherParent
  object Child2 {
    implicit val schema: Schema[Any, Child2] = Schema.gen
  }

  case class Child1(bool: Boolean) extends Parent with OtherParent
  object Child1 {
    implicit val schema: Schema[Any, Child1] = Schema.gen
  }

  case class Child3(str: String) extends Parent
  object Child3 {
    implicit val schema: Schema[Any, Child3] = Schema.gen
  }
}
```

Scala 2:
```graphql
schema {
  query: Queries
}

union Parent = Child1 | Child2 | Child3

type Child1 {
  bool: Boolean!
}

type Child2 {
  i: Int!
}

type Child3 {
  str: String!
}

type Queries {
  p: Parent!
}
```

Scala 3:
```graphql
schema {
  query: Queries
}

union Parent = Child1 | Child2 | Child3 | Child1 | Child2

type Child1 {
  bool: Boolean!
}

type Child2 {
  i: Int!
}

type Child3 {
  str: String!
}

type Queries {
  p: Parent!
}
```

### LINKED ISSUE #1993: Scala 3 different schema derivation nested enum items missing
Given the following:
```scala
sealed trait Enum1
object Enum1 {
  implicit val schema: Schema[Any, Enum1] = Schema.gen

  sealed trait Enum11 extends Enum1
  object Enum11 {
    implicit val schema: Schema[Any, Enum11] = Schema.gen

    case object Item111 extends Enum11 {
      implicit val schema: Schema[Any, Item111.type] = Schema.gen
    }

    case object Item112 extends Enum11 {
      implicit val schema: Schema[Any, Item112.type] = Schema.gen
    }
  }

  sealed trait Enum12 extends Enum1
  object Enum12 {
    implicit val schema: Schema[Any, Enum12] = Schema.gen

    case object Item121 extends Enum12 {
      implicit val schema: Schema[Any, Item121.type] = Schema.gen
    }
  }
}
```

Scala 2:
```graphql
schema {
  query: Queries
}

enum Enum1 {
  Item111
  Item112
  Item121
}

type Queries {
  e: Enum1!
}
```

Scala 3:
```graphql
schema {
  query: Queries
}

enum Enum1 {
  Enum11
  Enum12
}

type Queries {
  e: Enum1!
}
```

### PR DESCRIPTION
**Pull request by kyri-petrou**
Fixes #1992
Fixes #1993

... and hopefully all issues with nested sum types 🥲

### CODE REVIEW

File: core/src/main/scala-3/caliban/schema/SchemaDerivation.scala [hunk 1]
```diff
@@ -44,16 +44,22 @@ trait CommonSchemaDerivation {
...
-          (
-            constValue[name].toString,
-            MagnoliaMacro.anns[t], {
-              if (Macros.isEnumField[P, t])
-                if (!Macros.implicitExists[Schema[R, t]]) derived[R, t]
-                else summonInline[Schema[R, t]]
-              else summonInline[Schema[R, t]]
-            }.asInstanceOf[Schema[R, Any]]
-          ) :: values
+          inline summonInline[Mirror.Of[t]] match {
```
Comment (kyri-petrou) [inline core/src/main/scala-3/caliban/schema/SchemaDerivation.scala:47]: @ghostdogpr do you see any reason why we shouldn't do the same handling for `ArgBuilder`s? 

Tests all seem to pass with or without it but I'm wondering whether there are  any undiscovered edge-cases

Comment (kyri-petrou) [reply core/src/main/scala-3/caliban/schema/SchemaDerivation.scala:47]: OK after thinking about it I added it to the ArgBuilder as well since this is what matches Scala 2's / Magnolias behaviour.

Comment (ghostdogpr) [reply core/src/main/scala-3/caliban/schema/SchemaDerivation.scala:47]: Good call

**Review by ghostdogpr — APPROVED** [review]

**Comment by guymers** [comment]
My schema is equivalent between Scala 2 and 3 now, cheers!

Some of the unions aren't ordered alphabetically any more though.

**Comment by kyri-petrou** [comment]
Hey @guymers happy to hear it's all working as expected (well, for the most of it). We just [released a new version](https://github.com/ghostdogpr/caliban/releases/tag/v2.4.2?notification_referrer_id=NT_kwDOBALw57M4MzU0ODczMDI5OjY3MzAxNjA3) with all the fixes + some major performance improvements.

As for the ordering, if you can provide a repro and open an issue I can look into it.

And thanks again for reporting the issues, finally managed to get to the source of it and make the Scala 2 & 3 derivation equivalent (famous last words 😄 )

**Comment by kyri-petrou** [comment]
Hey @guymers, we've just discovered a regression that was introduced with these changes. Working on a fix now and we'll release a new version once I have the fix in, but in the meantime don't use version 2.4.2.

**Comment by guymers** [comment]
Only for Scala 3 yeah? I haven't merged cross version Scala 3 support yet for our GraphQL module and we wont move our app from 2.13 to 3 until after the holidays.

**Comment by kyri-petrou** [comment]
Yes only for Scala 3 👍 Should be fixed by #2005
